src/meta/src/hummock/hummock_manager.rs

```diff
@@ -455,88 +465,44 @@ where
     }
 
     /// Make sure `max_commited_epoch` is pinned and return it.
-    /// Assume that frontend will only pass the latest epoch value recorded by frontend to
-    /// `last_pinned`. Meta will unpin snapshots which are pinned and in (`last_pinned`,
-    /// `max_commited_epoch`).
-    pub async fn pin_snapshot(
-        &self,
-        context_id: HummockContextId,
-        last_pinned: HummockEpoch,
-    ) -> Result<HummockSnapshot> {
-        let mut versioning_guard = self.versioning.write().await;
-
-        // Use the max_committed_epoch in storage as the snapshot ts so only committed changes are
-        // visible in the snapshot.
-        let version_id = versioning_guard.current_version_id.id();
-
-        let max_committed_epoch = versioning_guard
-            .hummock_versions
-            .get(&version_id)
-            .unwrap()
-            .max_committed_epoch;
-
-        let mut pinned_snapshots = VarTransaction::new(&mut versioning_guard.pinned_snapshots);
+    pub async fn pin_snapshot(&self, context_id: HummockContextId) -> Result<HummockSnapshot> {
+        let max_committed_epoch = self.max_committed_epoch.load(Ordering::Relaxed);
+        let mut guard = self.versioning.write().await;
+        let mut pinned_snapshots = VarTransaction::new(&mut guard.pinned_snapshots);
         let mut context_pinned_snapshot = pinned_snapshots.new_entry_txn_or_default(
             context_id,
             HummockPinnedSnapshot {
                 context_id,
-                snapshot_id: vec![],
+                minimal_pinned_snapshot: max_committed_epoch,
             },
         );
-
-        // Unpin the snapshots pinned by meta but frontend doesn't know.
-        let to_unpin = context_pinned_snapshot
-            .snapshot_id
-            .iter()
-            .filter(|e| **e > last_pinned && **e < max_committed_epoch)
-            .cloned()
-            .collect_vec();
-        let mut snapshots_change = !to_unpin.is_empty();
-        for epoch in to_unpin {
-            context_pinned_snapshot.unpin_snapshot(epoch);
-        }
-
-        if !context_pinned_snapshot
-            .snapshot_id
-            .contains(&max_committed_epoch)
-        {
-            snapshots_change = true;
-            context_pinned_snapshot.pin_snapshot(max_committed_epoch);
+        if context_pinned_snapshot.minimal_pinned_snapshot == u64::MAX {
+            context_pinned_snapshot.minimal_pinned_snapshot = max_committed_epoch;
         }
 
-        if snapshots_change {
+        if context_pinned_snapshot.minimal_pinned_snapshot == max_committed_epoch {
             commit_multi_var!(self, Some(context_id), context_pinned_snapshot)?;
         }
```

Comment:
When will `context_pinned_snapshot.minimal_pinned_snapshot` be `u64::MAX`? IIUC, the correct logic is 
```
if context_pinned_snapshot.minimal_pinned_snapshot < max_committed_epoch {
  context_pinned_snapshot.minimal_pinned_snapshot = max_committed_epoch;
  commit_multi_var!(self, Some(context_id), context_pinned_snapshot)?;
}

Reply:
I use zero to present it is a just created item.